### ISSUE TITLE
Except for return/param rules, most rules should not be limited to functions

### ISSUE THREAD

**Issue by brettz9** [gajus/eslint-plugin-jsdoc#214]
Imo, except for return/param rules, most rules (e.g., rules checking types or tag names, e.g., `check-tag-names`) should not be limited to checking the jsdoc above functions, but wherever there are jsdoc comments (similar to the new `require-jsdoc` rule).

**Comment by golopot** [comment]
Some code examples that should be reported but are missed:

```js
/* eslint jsdoc/check-tag-names: 2 */

/** @typooo {number} */
let a = 4
```

```js
/* eslint jsdoc/valid-types: 2 */

/**
 * @typedef {!$#@!} .
 */
```

```js
/* eslint jsdoc/check-alignment: 2 */

/**
     * Hi
 */
class Foo {}
```

I suggest adding these checks in a major release, with no options or settings change.

**Comment by brettz9** [comment]
Fine by me, and if decided there is a need, we could add `contexts` later to override.

If this change is done, I think `match-description`, `require-jsdoc`, and `require-description` should default to the same contexts, though keeping the `contexts` override option, as descriptions are more stylistic and potentially customizable (unlike in your examples) and might be more likely to vary depending on context.

But note that there are a LOT of contexts, e.g., see https://github.com/typescript-eslint/typescript-eslint/blob/861844d727a2ba7f75cdbba2571498078aee2671/packages/typescript-estree/src/ts-estree/ast-node-types.ts , especially if you want to include typescript and JSX as well.

While PR #294 (inspired by #295) added `ObjectExpression` (and fixed `ClassExpression`), there may also be some need for special handling in `getJSDocComment`, e.g., as `FunctionExpression`, etc. check for comments not only immediately preceding, but in a few ancestors, e.g., if part of a variable declaration, the comment could be here:

```js
/**
*
*/
var a = function () {};
```

as well as:

```js
var a = /**
*
*/
function () {};
```

So there'd be a fair amount of work to analyze how those types ought to be checked and do it fairly systematically. Or, there is the option of just doing it as the need arises. Or something in the middle.
